### ISSUE TITLE
[Enhancement]: Avoid ELB resource deletion before ECS service definition update

### ISSUE THREAD

**Issue by at-wat** [hashicorp/terraform-provider-aws#34337]
### Description

This is mainly a problem in ECS/ELB API, but it would be nice if terraform-provider-aws could provide a workaround.

When deleting an ELB target group/listener/load balancer and unregistering the target group from an ECS service definition at once, no errors are returned, but ECS task deployments stall. (ELB resources can be removed before being unregistered from the ECS service definition)
Once entered into this situation, new task deployments won't start and will resume only after manually stopping all existing tasks, which causes service downtime.

It would be helpful if it's possible to control the order of ELB resource (target group/listener/load balancer) deletion and target group unregistering from ECS service definition.

### Affected Resource(s) and/or Data Source(s)

* aws_ecs_service
* aws_lb_listener
* aws_lb_target_group

### Potential Terraform Configuration

```terraform
# # How to reproduce
# 1. terraform init && terraform plan -out=plan && terrafrom apply
# 2. Remove code surrounded by vvvvvv and ^^^^^^ (remove and unregister LB at once)
# 3. terraform plan && terrafrom apply

locals {
  # Replace by your existing resources
  vpc_id = "**VPC_ID**"
  subnet_ids = [
    "**SUBNET1_ID**",
    "**SUBNET2_ID**",
  ]
  cluster_id = "arn:aws:ecs:**REGION**:**ACCOUNT_ID**:cluster/test-ecs-lb"
  task_definition_arn = "arn:aws:ecs:**REGION**:**ACCOUNT_ID**:task-definition/test-ecs-lb-svc-task:11"

  # # Requirements
  # task_definition_arn:
  #   - network mode must be "bridge"
  #   - container name must be "test-ecs-lb"
  # listener_port:
  #   - must be available on the LB
}

terraform {
  backend "local" {
  }
}

provider "aws" {
  region = "**REGION**"
}

resource "aws_lb" "main" {
  name               = "test-ecs-lb-bar"
  internal           = false
  subnets            = local.subnet_ids
  security_groups    = [aws_security_group.lb.id]
  load_balancer_type = "application"
}

resource "aws_security_group" "lb" {
  name   = "test-ecs-lb-bar"
  vpc_id = local.vpc_id

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = -1
    cidr_blocks = ["0.0.0.0/0"]
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_ecs_service" "main" {
  name            = "test-ecs-lb-bar"
  cluster         = local.cluster_id
  task_definition = local.task_definition_arn
  desired_count   = 1

  # vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
  #            Remove after the first apply
  load_balancer {
    target_group_arn = aws_lb_target_group.main.arn
    container_name   = "test-ecs-lb"
    container_port   = 80
  }
  # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  lifecycle {
    ignore_changes = [desired_count]
  }
}

# vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
#            Remove after the first apply
resource "aws_lb_listener" "main0" {
  load_balancer_arn = aws_lb.main.arn
  port              = "80"
  protocol          = "HTTP"

  default_action {
    type             = "forward"
    target_group_arn = aws_lb_target_group.main.arn
  }
}

resource "aws_lb_target_group" "main" {
  name        = "test-ecs-lb-bar-tg"
  vpc_id      = local.vpc_id
  protocol    = "HTTP"
  port        = 80
  target_type = "instance"

  deregistration_delay = 30

  health_check {
    path = "/"
  }

  lifecycle {
    create_before_destroy = true
  }
}
# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

### References

- https://github.com/aws/containers-roadmap/issues/2196

### Would you like to implement a fix?

None

**Comment by saiya** [comment]
As a workaround, following would be helpful?

- Add `depends_on` to the `aws_lb_target_group` in the `aws_ecs_service`
- `ecs_service.wait_for_steady_state` = true

By doing all above, I expect terraform updates ECS service (to unlink ALB target group from it) and awaits ECS tasks deployment. Then, deletes target group after completion of it (not in parallel with the ECS service/task update).
